packages/core/src/PartialExecutionUtils/findStartNodes.ts

```diff
@@ -0,0 +1,133 @@
+import type { INode, IPinData, IRunData } from 'n8n-workflow';
+import type { DirectedGraph } from './DirectedGraph';
+import { getIncomingData } from './getIncomingData';
+
+// TODO: implement dirty checking for options and properties and parent nodes
+// being disabled
+export function isDirty(node: INode, runData: IRunData = {}, pinData: IPinData = {}): boolean {
+	//- it’s properties or options changed since last execution, or
+
+	const propertiesOrOptionsChanged = false;
+
+	if (propertiesOrOptionsChanged) {
+		return true;
+	}
+
+	const parentNodeGotDisabled = false;
+
+	if (parentNodeGotDisabled) {
+		return true;
+	}
+
+	//- it has an error, or
+
+	const hasAnError = false;
+
+	if (hasAnError) {
+		return true;
+	}
+
+	//- it does neither have run data nor pinned data
+
+	const hasPinnedData = pinData[node.name] !== undefined;
+
+	if (hasPinnedData) {
+		return false;
+	}
+
+	const hasRunData = runData?.[node.name];
+
+	if (hasRunData) {
+		return false;
+	}
+
+	return true;
+}
+
+function findStartNodesRecursive(
+	graph: DirectedGraph,
+	current: INode,
+	destination: INode,
+	runData: IRunData,
+	pinData: IPinData,
+	startNodes: Set<INode>,
+	seen: Set<INode>,
+): Set<INode> {
+	const nodeIsDirty = isDirty(current, runData, pinData);
+
+	// If the current node is dirty stop following this branch, we found a start
+	// node.
+	if (nodeIsDirty) {
+		startNodes.add(current);
+
+		return startNodes;
+	}
+
+	// If the current node is the destination node stop following this branch, we
+	// found a start node.
+	if (current === destination) {
+		startNodes.add(current);
+		return startNodes;
+	}
+
+	// If we detect a cycle stop following the branch, there is no start node on
+	// this branch.
+	if (seen.has(current)) {
+		return startNodes;
+	}
+
+	// Recurse with every direct child that is part of the sub graph.
+	const outGoingConnections = graph.getDirectChildren(current);
+	for (const outGoingConnection of outGoingConnections) {
+		const nodeRunData = getIncomingData(
```

Comment:
I agree. And for partial executions it's probably a good idea to do that, because it's not performance sensitive. Nothing here runs in a tight loop.

I hope that we won't do this in the execution engine as well though, so we don't create a lot of short lived objects that trigger smaller GCs. 

Can I make that change in a follow up PR? I created an issue for this: https://linear.app/n8n/issue/PAY-1940/use-option-objects-for-functions-with-a-lot-of-parameters

Reply:
Yes perfectly fine as a follow-up 👌 In general I don't mind if we fix certain things in separate follow-up PRs. Just as long as we don't forget it :)